Hugging Face Transformer Deployment Tutorial (#49)
triton-inference-server/tutorials: Quick_Deploy/HuggingFaceTransformers/README.md

These models were selected because of their small size and consistent response quality, 
enabling the tutorial to run on most devices with sensible results. However, this tutorial
is also generalizable for much larger models provided sufficient infrastructure. 

*NOTE*: The tutorial is intended to be a reference example only. It is a work in progress.

## Step 1: Build a Model Repository

The first step is to create a model repository that the Triton Inference Server will use
for inference processing. We can create a model repository from the provided base files by 
executing the following python script:

```
python3 create_repository.py
```

Without specifying any arguments, this script will create a model repository containing
the details necessary for Triton to load and serve each of the aforementioned models.
For the purposes of this tutorial, we provide Triton with simple ```config.pbtxt``` files
for each model that describe:
- The the backend to use.
- Model input and output details.
- Custom parameters to use for execution.

> ## Step 2: Build a Triton Container Image

Reviewer: I'm confused about the best practice in shipping a custom Python execution environment, spawning a new image, or using [conda-pack](https://github.com/triton-inference-server/python_backend#creating-custom-execution-environments).

What do you prefer in production-level deployment?

Reviewer 2: @Tabrizian  - thoughts on this?

Reviewer 3: That depends on your goals. If you want to run models with different Python environments on one server, you would [use custom execution environments](https://github.com/triton-inference-server/python_backend#creating-custom-execution-environments). It's more memory-intensive (especially as you increase the number of models), so if you only need one Python version and environment, it would be easier to use an image with all the components needed.

Also, conda-pack can reduce the size of the environment itself (versus using an uncompressed environment), but it then needs to be unpacked at load time, which increases the model load time considerably.

Reviewer 4: Adding to what David mentioned, if you don't want to create a new container that has all the dependencies or you want to isolate the dependencies used by each model custom execution environments are the recommended path.

The second step is to create an image that includes all the dependencies necessary
to deploy hugging face transformer models on the Triton Inference Server. This can be done
by building an image from the provided Dockerfile.

```
docker build -t triton_transformer_server .
```

## Step 3: Run the Triton Inference Server

Once the ```triton_transformer_server``` image is created, you can run a container with
the following command:

```bash
docker run --gpus all -it --rm -p 8000:8000 --shm-size=1G --ulimit memlock=-1 --ulimit stack=67108864 -v ${PWD}/model_repository:/opt/tritonserver/model_repository triton_transformer_server tritonserver --model-repository=model_repository
```

The server has launched successfully when you see the following outputs in your console:

```bash
I0922 23:28:40.351809 1 grpc_server.cc:2451] Started GRPCInferenceService at 0.0.0.0:8001
I0922 23:28:40.352017 1 http_server.cc:3558] Started HTTPService at 0.0.0.0:8000
I0922 23:28:40.395611 1 http_server.cc:187] Started Metrics Service at 0.0.0.0:8002
```

## Step 4: Use a Triton Client to Query the Server

Now we can query the server using the client script provided.

```bash

> python3 client.py -m [model_name] --prompt [prompt_message]

Reviewer 2: can we add curl as reference?

Reviewer 5: This would be blocked until Guan and I finish implementing simpler curl format support, unless you mean a tensor-oriented curl:
```
curl -X POST localhost:8000/v2/models/text/infer -d '{"inputs": [{"name":"prompt","datatype":"BYTES","shape":[1,1],"data":["hello"]}]}'
```

Reviewer 2: I think the simple curl one liner is good for now - and we will add the generate version when available

# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-->

# Deploying Hugging Face Transformer Models in Triton

The following tutorial demonstrates how to deploy an arbitrary hugging face transformer
model on the Triton Inference Server using Triton's [Python backend](https://github.com/triton-inference-server/python_backend). For the purposes of this example, three transformer
models will be deployed:
- [tiiuae/falcon-7b](https://huggingface.co/tiiuae/falcon-7b)
- [adept/persimmon-8b-base](https://huggingface.co/adept/persimmon-8b-base)
- [mistralai/Mistral-7B-v0.1](https://huggingface.co/mistralai/Mistral-7B-v0.1)

These models were selected because of their popularity and consistent response quality.
However, this tutorial is also generalizable for any transformer model provided 
sufficient infrastructure. 

> *NOTE*: The tutorial is intended to be a reference example only. It is a work in progress.

Reviewer 5: ```suggestion
*NOTE*: The tutorial is intended to be a reference example only. It may not be tuned for optimal performance.
```

Reviewer 5: Calling it a WIP may deter users from trying it, thinking it may not work.

Author: I agree. In that same vein then do we want to change the the WIP phrase for the vLLM [tutorial](https://github.com/triton-inference-server/tutorials/blob/512ba538fd81b37ad0cc3133a6d7905f72a9ad04/Quick_Deploy/vLLM/README.md?plain=1#L37C66-L37C91) to something like: 
"The tutorial is intended to be a reference example only and has known limitations."?

Reviewer 5: sounds good to me CC @tanmayv25

Author: I went ahead and changed it. I can revert it if need be.

## Step 1: Create a Model Repository

The first step is to create a model repository containing the models we want the Triton 
Inference Server to load and use for inference processing. For this tutorial an empty
directory named `model_repository` has been provided, into which we will copy the 

> `falcon7b` model folder:

Reviewer 5: Empty folder won't work on git. Just update instructions to `mkdir -p model_repository` first.

Author: TIL

## Step 2: Build a Triton Container Image

The second step is to create an image that includes all the dependencies necessary
to deploy hugging face transformer models on the Triton Inference Server. This can be done
by building an image from the provided Dockerfile:

```
docker build -t triton_transformer_server .
```

## Step 3: Launch the Triton Inference Server

Once the ```triton_transformer_server``` image is created, you can launch the Triton Inference
Server in a container with the following command:

```bash
docker run --gpus all -it --rm -p 8000:8000 --shm-size=1G --ulimit memlock=-1 --ulimit stack=67108864 -v ${PWD}/model_repository:/opt/tritonserver/model_repository triton_transformer_server tritonserver --model-repository=model_repository
```

The server has launched successfully when you see the following outputs in your console:

```
I0922 23:28:40.351809 1 grpc_server.cc:2451] Started GRPCInferenceService at 0.0.0.0:8001
I0922 23:28:40.352017 1 http_server.cc:3558] Started HTTPService at 0.0.0.0:8000
I0922 23:28:40.395611 1 http_server.cc:187] Started Metrics Service at 0.0.0.0:8002
```

> ## Step 4: Use a Triton Client to Query the Server

Reviewer 5: ```suggestion
## Step 4: Query the Server
```

that the server has launched successfully.

Query the server making sure to change the host address for each model:
```json
curl -X POST localhost:8000/v2/models/falcon7b/infer -d '{"inputs": [{"name":"prompt","datatype":"BYTES","shape":[1],"data":["How can you be"]}]}'
curl -X POST localhost:8000/v2/models/mistral7b/infer -d '{"inputs": [{"name":"prompt","datatype":"BYTES","shape":[1],"data":["Where are you going"]}]}'
curl -X POST localhost:8000/v2/models/persimmon8b/infer -d '{"inputs": [{"name":"prompt","datatype":"BYTES","shape":[1],"data":["Where is the nearest"]}]}'
```
In our testing, these queries return the following parsed results:
```bash
# falcon7b
"How can you be sure that you are getting the best deal on your car"

# mistral7b
"Where are you going? I’m going to the beach."

# persimmon8b
"Where is the nearest starbucks?"
```
## 'Hour Zero' Support

At the time of writing this tutorial, transformers version 4.34.0 was not yet released, meaning
very new models such as Persimmon-8B and Mistral 7B were not yet fully supported by the latest
transformers releases (4.33.3). Triton is not limited to waiting for official releases and can

> load the latest models by building from source. This can be done by replacing:

Reviewer 5: We may want to avoid specific versions and models here to avoid dating it too much, as these models mentioned will quickly become "old" compared to newer ones. We should instead just say something more generic about how bleeding edge / latest models may not be supported in the latest published release of `transformers`, so for bleeding edge model support you can install latest transformers from source. To do so, replace the transformers install in the `Dockerfile` with `pip install git+https://github.com/huggingface/transformers.git` ...

Using this technique you should be able to serve any models supported by HuggingFace Transformer text generation pipeline with Triton.

---

above is just the gist, feel free to rephrase, just want to avoid mention of specific versions and models in this disclaimer

Author: Great point!

model or restart it using the docker run command from above.

Once Triton launches successfully, start a Triton SDK container by running the following in a separate window:

```bash
docker run -it --net=host nvcr.io/nvidia/tritonserver:23.09-py3-sdk bash
```
This container comes with all of Triton's deployment analyzers pre-installed, meaning
we can simply enter the following to get feedback on our model's inference performance:

```bash
perf_analyzer -m falcon7b --concurrency-range 1:4
```

This command should run quickly and profile the performance of our falcon7b model at
increasing levels of concurrency. As the analyzer runs, it will output useful metrics
such as latency percentiles, latency by stage of inference, and successful request
count. Ultimately, the analyzer will neatly summarize the data in the final output:

```json
Concurrency: 1, throughput: 23.2174 infer/sec, latency 43041 usec
Concurrency: 2, throughput: 23.3284 infer/sec, latency 85590 usec
Concurrency: 3, throughput: 23.94 infer/sec, latency 125085 usec
Concurrency: 4, throughput: 23.773 infer/sec, latency 167879 usec

> ```

Reviewer 5: Not sure if this is specific feedback, but since the models aren't optimized for batching, the throughput stays flat here and latency gets worse linearly with concurrency - so I don't think we should show concurrency > 1 for now.

Author: We were asked to incorporate tangible ways to measure performance and optimize a generic deployment. Seemed like a great time to point to the tools team. But yes, this is a strong point! For now, I've changed it to just collect metrics with concurrency 1. The full detail list the perf analyzer produces is this:
```Request concurrency: 1
  Client:
    Request count: 422
    Throughput: 23.4404 infer/sec
    Avg latency: 42583 usec (standard deviation 3809 usec)
    p50 latency: 41547 usec
    p90 latency: 43611 usec
    p95 latency: 48919 usec
    p99 latency: 61778 usec
    Avg HTTP time: 42576 usec (send/recv 87 usec + response wait 42489 usec)
  Server:
    Inference count: 422
    Execution count: 422
    Successful request count: 422
    Avg request latency: 42140 usec (overhead 24 usec + queue 26 usec + compute input 26 usec + compute infer 41998 usec + compute output 65 usec)

  Server Prometheus Metrics:
    Avg GPU Utilization:
      GPU-57c7b00e-ca04-3876-91e2-c1eae40a0733 : 52.7222%
    Avg GPU Power Usage:
      GPU-57c7b00e-ca04-3876-91e2-c1eae40a0733 : 234.525 watts
    Max GPU Memory Usage:
      GPU-57c7b00e-ca04-3876-91e2-c1eae40a0733 : 35431383040 bytes
    Total GPU Memory:
      GPU-57c7b00e-ca04-3876-91e2-c1eae40a0733 : 51527024640 bytes
Inferences/Second vs. Client Average Batch Latency
Concurrency: 1, throughput: 23.4404 infer/sec, latency 42583 usec```

If you think any of these metrics would be more relevant than the ones I selected, let me know and I can swap them out!

Reviewer 5: Looks great! I haven't actually seen the GPU metric collection feature being used until now, very cool!

Reviewer 5: If you have spare cycles for a couple more runs - I notice the GPU utilization is ~50%. I wonder if we can hit 100% with 2+ instances of the model (if it can fit, I see memory usage is > half).

Reviewer 5: (Disclaimer) This and the previous comment are totally separate from this tutorial PR and only if you have spare cycles/curiosity.

If simply scaling instances doesn't work, another thing to try eventually is batching. Long term we probably want to move away from pipeline to have more fine tuned control over batching, but for a quick example, you could enable dynamic batching on the model, either set high concurrency on perf analyzer or set high max_queue_delay in config.pbtxt to allow larger batch to be gathered, and I think you can do something like this to [batch with the pipeline](https://huggingface.co/docs/transformers/main_classes/pipelines#pipeline-batching):
```python
prompts = []
for request in requsts:
   prompts.append(prompt)

# or some fixed batch size? 
batch_size = len(prompts)
sequences = self.pipeline(prompts, batch_size=batch_size, ...)

responses = []
for seq in sequences:
  # tensor = ...
  # response = ...
  responses.append(response)

return responses
```

Author: As a matter of fact, I know it can hit ~100%, because when I was building out the model analyzer part of the tutorial it recommended a config with 2 GPU instances for that very reason. Maybe better to show the output in the MA section to show this very value.

to deploy hugging face transformer models on the Triton Inference Server. This can be done
by building an image from the provided Dockerfile:

```
docker build -t triton_transformer_server .
```

## Step 3: Launch the Triton Inference Server

Once the ```triton_transformer_server``` image is created, you can launch the Triton Inference
Server in a container with the following command:

```bash
docker run --gpus all -it --rm --net=host --shm-size=1G --ulimit memlock=-1 --ulimit stack=67108864 -v ${PWD}/model_repository:/opt/tritonserver/model_repository triton_transformer_server tritonserver --model-repository=model_repository
```

The server has launched successfully when you see the following outputs in your console:

```
I0922 23:28:40.351809 1 grpc_server.cc:2451] Started GRPCInferenceService at 0.0.0.0:8001
I0922 23:28:40.352017 1 http_server.cc:3558] Started HTTPService at 0.0.0.0:8000
I0922 23:28:40.395611 1 http_server.cc:187] Started Metrics Service at 0.0.0.0:8002
```

## Step 4: Query the Server

Now we can query the server using curl, specifying the server address and input details:

```json

> curl -X POST localhost:8000/v2/models/falcon7b/infer -d '{"inputs": [{"name":"text_input","datatype":"BYTES","shape":[1],"data":["I am going"]}]}'

Reviewer 5: We can also curl with the `/generate` endpoint now. Feel free to add in a separate PR if you want though.

Reviewer 5: ref: https://github.com/triton-inference-server/server/blob/main/docs/protocol/extension_generate.md

Author: I'll create a minor ticket for this and wait for the 23.10 container to be released.

Dockerfile with:
```docker
RUN pip install git+https://github.com/huggingface/transformers.git
```
Using this technique you should be able to serve any transformer models supported by
hugging face with Triton.

# Next Steps
The following sections expand on the base tutorial and provide guidance for future sandboxing.

## Loading Cached Models
In the previous steps, we downloaded the falcon-7b model from hugging face when we
launched the Triton server. We can avoid this lengthy download process  in subsequent runs
by loading cached models into Triton. To do so, we can follow the hugging face [tutorial
for downloading models](https://huggingface.co/docs/hub/models-downloading). Once the model
is downloaded, we can mount it to the Triton container by adding the following mount option to our
`docker run` command from earlier (making sure to replace {USER} with your username on
Linux):

```bash
# Option to mount a specific cached model (falcon-7b in this case)
-v /home/{USER}/.cache/huggingface/hub/models--tiiuae--falcon-7b:/root/.cache/huggingface/hub/models--tiiuae--falcon-7b

# Option to mount all cached models on the host system

> -v /home/{USER}/.cache/huggingface:/root/.cache/huggingface

Reviewer 5: This is a great addition! Do you know if the cache directory is configurable for HF? Would be cool if we could eventually set it to cache where the model.py lives / in the model repository or something.

Author: Indeed it is, and indeed we can. I'll extend this section.

> -v /home/{USER}/.cache/huggingface:/root/.cache/huggingface

Reviewer 5: `/home/${USER}` can be `${HOME}` instead for brevity

cp -r persimmon8b/ model_repository/
```
*NOTE*: The combined size of these three models is large. If your current hardware cannot
support hosting all three models simultaneously, consider copying only a single additional
model.

Again, launch the server by invoking the `docker run` command from above and wait for confirmation
that the server has launched successfully.

Query the server making sure to change the host address for each model:
```json
curl -X POST localhost:8000/v2/models/falcon7b/infer -d '{"inputs": [{"name":"prompt","datatype":"BYTES","shape":[1],"data":["How can you be"]}]}'
curl -X POST localhost:8000/v2/models/mistral7b/infer -d '{"inputs": [{"name":"prompt","datatype":"BYTES","shape":[1],"data":["Where are you going"]}]}'
curl -X POST localhost:8000/v2/models/persimmon8b/infer -d '{"inputs": [{"name":"prompt","datatype":"BYTES","shape":[1],"data":["Where is the nearest"]}]}'
```
In our testing, these queries return the following parsed results:
```bash
# falcon7b
"How can you be sure that you are getting the best deal on your car"

# mistral7b
"Where are you going? I’m going to the beach."

# persimmon8b
"Where is the nearest starbucks?"
```

> ## 'Hour Zero' Support

Reviewer 5: "Hour zero" was just me being cheeky. I think we should probably call it "day zero" or something more generic like "bleeding edge models" or something. CC @nnshah1 who probably has some thoughts

Author: I appreciated the cheekiness, but I guess we can change it to something more tame... :disappointed:
